SHA3.cs

```diff
@@ -0,0 +1,717 @@
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static bool TryHashData(
+        ReadOnlySpan<byte> key,
+        ReadOnlySpan<byte> source,
+        Span<byte> destination,
+        out int bytesWritten);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static byte[] HashData(byte[] key, Stream source);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static byte[] HashData(ReadOnlySpan<byte> key, Stream source);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static int HashData(ReadOnlySpan<byte> key, Stream source, Span<byte> destination);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static ValueTask<byte[]> HashDataAsync(byte[] key, Stream source, CancellationToken cancellationToken = default);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static ValueTask<byte[]> HashDataAsync(ReadOnlyMemory<byte> key, Stream source, CancellationToken cancellationToken = default);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static ValueTask<int> HashDataAsync(
+        ReadOnlyMemory<byte> key,
+        Stream source,
+        Memory<byte> destination,
+        CancellationToken cancellationToken = default);
+}
+
+public abstract partial class Shake : IDisposable {
+    internal Shake(); // Will not be subclassable by developers.
+
+    public void AppendData(byte[] data);
+    public void AppendData(ReadOnlySpan<byte> data);
+    protected abstract void AppendDataCore(ReadOnlySpan<byte> data);
+
+    public byte[] GetCurrentHash(int outputLength);
+    public void GetCurrentHash(Span<byte> destination);
+    protected abstract void GetCurrentHashCore(Span<byte> destination);
+
+    public byte[] GetHashAndReset(int outputLength);
+    public void GetHashAndReset(Span<byte> destination);
+    protected abstract void GetHashAndResetCore(Span<byte> destination);
+
+    public void Dispose();
+    protected virtual void Dispose(bool disposing);
+}
+
+public sealed partial class Shake128 : Shake {
+
+    [UnsupportedOSPlatform("ios")]
```

Comment:
In theory since the SHAKE classes are sealed and have no static or const members that are useful, we could mark the whole class as UOSP. The guard attributes on `IsSupported` would still do the right thing.